lib/seahorn/SolverBmc.cc

```diff
@@ -0,0 +1,449 @@
+#include "seahorn/SolverBmc.hh"
+#include "seahorn/Transforms/Instrumentation/ShadowMemDsa.hh"
+#include "seahorn/UfoOpSem.hh"
+
+#include "llvm/IR/DebugInfo.h"
+#include "llvm/IR/DebugLoc.h"
+#include "llvm/Support/CommandLine.h"
+
+#include "boost/container/flat_set.hpp"
+#include "seahorn/Expr/ExprLlvm.hh"
+#include "seahorn/Expr/HexDump.hh"
+#include "seahorn/Support/SeaDebug.h"
+
+#ifdef WITH_YICES2
+#include "seahorn/Expr/Smt/Yices2SolverImpl.hh"
+#endif
+
+static llvm::cl::opt<bool>
+    SDumpHex("horn-bmc-solver-hexdump",
+             llvm::cl::desc("Dump memory state using hexdump"),
+             cl::init(false));
+static llvm::cl::opt<std::string>
+    SBmcSmtTactic("horn-bmc-solver-tactic",
+                  llvm::cl::desc("Z3 tactic to use for BMC"),
+                  cl::init("default"));
+static llvm::cl::opt<std::string>
+    SBmcSmtLogic("horn-bmc-solver-logic",
+                 llvm::cl::desc("SMT-LIB logic to pass to Z3"),
+                 cl::init("ALL"));
+
+namespace seahorn {
+SolverBmcEngine::SolverBmcEngine(OperationalSemantics &sem, EZ3 &zctx,
+                                 solver::SolverKind solver_kind)
+    : m_sem(sem), m_efac(sem.efac()), m_result(solver::SolverResult::UNKNOWN),
+      m_cpg(nullptr), m_fn(nullptr), m_ctxState(m_efac),
+      m_solver_kind(solver_kind) {
+
```

Comment:
remove z3 specific stuff. we will need solver independent way of specifying options if needed

Reply:
do we still need z3 specific stuff for the z3 solver used in calculating unsat core?

Reply:
some functionality (simplification, incremental solving) is z3 specific and we can still use z3 for that.
Some functionality is z3 specific but is only used for debug purposes and can be removed and then rewritten using different interfaces (unsat core)
Some functionality is z3 specific but will need to be ported to more generic interface for full functionality (implicant computation and counterexamples)

Reply:
don't worry about creating additional EZ3 context if they are needed